### DISCUSSION TITLE
There seem to be memory leaks with psycopg 3 (3.1.9)

### DISCUSSION THREAD

**Discussion by nmoreaud** [sqlalchemy/sqlalchemy#10270] [Usage Questions] [answered]
Hello

There seem to be memory leaks with `psycopg3[binary]` that were not present with `psycopg2-binary` when doing multiple selects.
I have this kind of tracemalloc:
```
"<string>:1: size=9000 KiB (+8284 KiB), count=127393 (+117810), average=72 B
/project/.venv/lib/python3.8/site-packages/psycopg/_queries.py:355: size=7353 KiB (+6854 KiB), count=126318 (+117757), average=60 B
/project/.venv/lib/python3.8/site-packages/psycopg/_queries.py:306: size=5459 KiB (+5073 KiB), count=126318 (+117757), average=44 B
/project/.venv/lib/python3.8/site-packages/psycopg/_queries.py:63: size=3200 KiB (+2960 KiB), count=69 (+42), average=46.4 KiB
/project/.venv/lib/python3.8/site-packages/psycopg/_queries.py:195: size=1964 KiB (+1813 KiB), count=133 (+84), average=14.8 KiB
/project/.venv/lib/python3.8/site-packages/psycopg/_queries.py:184: size=1042 KiB (+971 KiB), count=63 (+42), average=16.5 KiB
/project/.venv/lib/python3.8/site-packages/psycopg/_queries.py:182: size=1042 KiB (+971 KiB), count=63 (+42), average=16.5 KiB
/project/.venv/lib/python3.8/site-packages/psycopg/_queries.py:365: size=1042 KiB (+971 KiB), count=61 (+42), average=17.1 KiB
/usr/lib/python3.8/tracemalloc.py:532: size=122 KiB (+122 KiB), count=2212 (+2211), average=56 B
/project/.venv/lib/python3.8/site-packages/psycopg/_queries.py:75: size=10.2 KiB (+5800 B), count=140 (+84), average=75 B
/project/.venv/lib/python3.8/site-packages/psycopg/cursor.py:864: size=535 KiB (+5269 B), count=7935 (+75), average=69 B
/project/.venv/lib/python3.8/site-packages/psycopg/_queries.py:314: size=3920 B (+2352 B), count=70 (+42), average=56 B
/project/.venv/lib/python3.8/site-packages/psycopg/_queries.py:164: size=3920 B (+2352 B), count=70 (+42), average=56 B
/project/.venv/lib/python3.8/site-packages/psycopg/_queries.py:175: size=3528 B (+2352 B), count=63 (+42), average=56 B
/project/.venv/lib/python3.8/site-packages/psycopg/_queries.py:310: size=3465 B (+1806 B), count=63 (+42), average=55 B
/project/.venv/lib/python3.8/site-packages/sqlalchemy/sql/cache_key.py:280: size=4808 B (+1712 B), count=9 (+3), average=534 B
/project/.venv/lib/python3.8/site-packages/psycopg/_adapters_map.py:76: size=9976 B (+1512 B), count=60 (+9), average=166 B
/project/.venv/lib/python3.8/site-packages/sqlalchemy/sql/default_comparator.py:51: size=1461 B (+1461 B), count=2 (+2), average=730 B
/usr/lib/python3.8/tracemalloc.py:397: size=728 B (+728 B), count=4 (+4), average=182 B
/project/.venv/lib/python3.8/site-packages/psycopg/_adapters_map.py:75: size=9992 B (+672 B), count=66 (+4), average=151 B"
```

Example:
```python
while True:
    for i in range(50):
        load()
    gc.collect()
    time.sleep(5)
    # look at htop => each time memory increases by 10MB.

def load():
    q = cnx.session.query(Table1.col1, Table1.col2, Table1.rel_name)
    q = q.limit(50000)
    rel_names = {t.rel_name for t in q.all()}
    
    q = (
        session.query(Table2.col1, Table2.col2, ...)
        .filter(Table2.name.in_(rel_names))
    )
    q.all()
    # session.commit()
    # session.rollback()
```

I don't know if it comes from psycopg3, SQLAlchemy or if it is a normal behavior.

**Comment by CaselIT** [comment]
Hi,
Does adding a rollback / commit resolvers the issue?

Btw if a leak exists it seems either in psycopg or in how sqlalchemy uses psycopg

**Reply by nmoreaud** [reply]
No the issue remains

**Reply by CaselIT** [reply]
since the code above is not exactable, does the type of query or size matter?

**Reply by nmoreaud** [reply]
There is something really strange: the first select never returns result in the same order.
Each time this select is emitted, the rows order in the table is modified (I check with dbeaver). But there are no inserts/update??

**Reply by CaselIT** [reply]
a select without orderby doesn't have an order, it can return rows in any order. You can think of a table as a set of row. If you don't sort it a set will iterate in "random" order

**Reply by nmoreaud** [reply, marked as answer]
Okay, so this is it: I've been bitten by postgresql `synchronize_seqscans`.
And psycopg3 uses a lot more memory than psycopg2 when executing multiple `select in` queries with a lot of elements (a different list each time). It probably uses a cache, so not really a memory leak (it tends to a fixed limit)

![image](https://github.com/sqlalchemy/sqlalchemy/assets/2413781/63ae902d-6a1c-46e2-8b4e-ca4428e668a3)

**Reply by CaselIT** [reply]
since this seems a psycopg issue, like mike mentioned you should report it to their repo

**Comment by zzzeek** [comment]
can you take this to https://github.com/psycopg/psycopg and provide reproducing information?    If a reproducing example isn't available, there's nothing we can do here and it's not clear if @dvarrazzo could do much either

**Reply by nmoreaud** [reply]
Copied at: https://github.com/psycopg/psycopg/discussions/628

**Reply by CaselIT** [reply]
if sqlalchemy needs to do something different feel free to update this discussion or tag me or mike in the psycopg issue

**Comment by dvarrazzo** [comment]
@CaselIT, @zzzeek: please correct me if I'm wrong:

```
    q = (
        session.query(Table2.col1, Table2.col2, ...)
        .filter(Table2.name.in_(rel_names))
    )
```

I assume that, in this query, SA creates a big query with an hardcoded list of the values taken from `rel_names`, instead of using parameters.

If that's the case, the user is finding an `@lru_cache(maxsize=128)` where every single query is a different entry, and, because the names lists are thousands of elements, each entry is chubby.

https://github.com/psycopg/psycopg/blob/b8f27fc81be8e681a1cf11c36926dc16f98c7243/psycopg/psycopg/_queries.py#L147

Can the user use instead something that generates `WHERE table2.name = any(%s)` instead, with the list of values passed as argument? This would benefit pretty much the whole pipeline (faster/cached parsing on the client, faster parsing on the server, prepared statements would kick in).

**Reply by nmoreaud** [reply]
My application is compatible with multiple database vendors (Postgres, Oracle, MSSQL, etc), so I cannot adapt the functions used for each vendor. If IN and ANY are 100% equivalent, SQLAlchemy internals could convert to ANY for optimized driver support.
However, I can batch `where in` by 500 like `selectinload` sqlalchemy function does, which seems to work fine. 
The memory increase seems to occur only when `where in` clause size changes.

**Reply by dvarrazzo** [reply]
I see. Probably then SA creates different queries `WHERE field in (%s, %s, ...)` with varying number of placeholders; each one gets cached as a separate entry by the query parser LRU cache.

**Reply by CaselIT** [reply]
> Can the user use instead something that generates WHERE table2.name = any(%s) instead, with the list of values passed as argument? This would benefit pretty much the whole pipeline (faster/cached parsing on the client, faster parsing on the server, prepared statements would kick in).

this is not currently supported by sqlalchemy since dbapi/drivers usually don't support `in (<single_param>)` with `single_param=[1,2,...,n]`.
I think this change could be implemented for pg if all driver (maybe this isn't required) we support have this ability (I know psycopg2 also support this). I'm assuming that if `single_param` is empty this would not be handled by the driver though. @zzzeek what's your opinion on this?

> I assume that, in this query, SA creates a big query with an hardcoded list of the values taken from rel_names, instead of using parameters.

parameters are used but they are broken up one for each value. so when using an in with 10 values, the query would have 10 parameters written inside the `(%(p1)s, %(p2)s, ..., %(p10)s)`

here is an example from the test (for the default dialect, not for pg) https://github.com/sqlalchemy/sqlalchemy/blob/5f0c591c744ddf17cf5d886aec21e242eb58562f/test/sql/test_compiler.py#L4854-L4866

@dvarrazzo is it possible to disable that caching by the user of psyocopg (something like `execute(..., cache=False)`)? Maybe a temporary solution could be that either sqlalchemy disables completely the psycopg query cache (I guess it would not be great) or just when a statement with an IN is executed (harder to do, but I believe it should be doable since sqlalchemy knows that it needs to render the IN before executing)

**Reply by dvarrazzo** [reply]
@CaselIT The cache is an implementation detail, I don't really desire to make it visible to the user.

Postgres has a better construct than `IN` (which doesn't support an empty list of parameters) and, because Postgres native placeholders don't support `IN %s` anyway, [we suggest](https://www.psycopg.org/psycopg3/docs/basic/from_pg2.html#you-cannot-use-in-s-with-a-tuple) to use `= any(%s)`, which works for both psycopg2, 3, aiopg, pg8000. Frankly I don't wish to dumb-down psycopg because other database vendors don't have the same features that Postgres has. 

@nmoreaud: 

> My application is compatible with multiple database vendors (Postgres, Oracle, MSSQL, etc), so I cannot adapt the functions used for each vendor

Actually I am pretty sure that you can, and it would be pretty simple to do that, by checking which connection you are using.

**Reply by CaselIT** [reply]
> The cache is an implementation detail, I don't really desire to make it visible to the user

Understandable. But maybe a global "off" could be something useful?

> Postgres has a better construct than IN (which doesn't support an empty list of parameters) and, because Postgres native placeholders don't support IN %s anyway, [we suggest](https://www.psycopg.org/psycopg3/docs/basic/from_pg2.html#you-cannot-use-in-s-with-a-tuple) to use = any(%s), which works for both psycopg2, 3, aiopg, pg8000. Frankly I don't wish to dumb-down psycopg because other database vendors don't have the same features that Postgres has.

Yes, I know about `=any()` and `!=all()` for `in/not in` and sqlalchemy supports them when rendered explicitly. What sqlalchemy currently does not is transform an in to any automatically.
Note that I'm not asking you to not support or suggest the use of any, since psycopg should care only about what's best for pg.


Not sure what would be best here, maybe a dialect parameter called something like "optimize_in_with_any" thwt would enable this change for postgresql? (I think it would not be that since since IIRC postcompile at the moment renders the `in/not in`)

Just to clarify, the lru is bounded, so the memory increases, but it's not unbounded, correct?

Let's wait mike opinion.

**Reply by dvarrazzo** [reply]
> Just to clarify, the lru is bounded, so the memory increases, but it's not unbounded, correct?

This is correct (if it was unbounded we would consider it no problem a genuine bug on our side).

```
>>> import psycopg
>>> conn = psycopg.connect("")
>>> for i in range(1000):
...     conn.execute(f"select {i}, %s", [1])

>>> psycopg._queries._query2pg.cache_info()
CacheInfo(hits=0, misses=1000, maxsize=128, currsize=128)
```

It is noticeable in the OP example because the size of each query is massive too, but the average query in a program, even if generated by an ORM, is in the order of the 100-1000 bytes, which would give around 1Mb of footprint. How large becomes a query with 5000 placeholders?

**Reply by CaselIT** [reply]
> How large becomes a query with 5000 placeholders?

well sqlalchemy sends `select ..the query.. where col in (%(param_x_1)s, %(param_x_2)s, ..., %(param_x_5000)s)` so that big

**Reply by nmoreaud** [reply]
In my examples, the textual queries take 12KB to 70KB (70 * 128 => 9MB) but the parsed query cache takes more than 100MB, there is an order of magnitude between the textual queries and psycopg parsed queries.

**Reply by zzzeek** [reply]
> I assume that, in this query, SA creates a big query with an hardcoded list of the values taken from rel_names, instead of using parameters.

no, SQLAlchemy uses bound parameters 100% of the time including for large IN lists

we have our own LRUCache in play for statements, however for large IN lists we abstract this into a "parameter" of our own that is late-rendered so that these varying-size lists don't create additional entries

> Postgres has a better construct than IN (which doesn't support an empty list of parameters)

our IN supports empty lists using same mechanism above, we late-render a special "empty list" evaluation when the list is empty :)

> Can the user use instead something that generates `WHERE table2.name = any(%s)` instead, with the list of values passed as argument? This would benefit pretty much the whole pipeline (faster/cached parsing on the client, faster parsing on the server, prepared statements would kick in).

> Not sure what would be best here, maybe a dialect parameter called something like "optimize_in_with_any" thwt would enable this change for postgresql? (I think it would not be that since since IIRC postcompile at the moment renders the in/not in)

this is technically possible, I definitely would not want to just "switch it on" for all drivers, at the same time this optimization could be part of the base PostgreSQL dialect since it uses a PG syntax.   However, the slightest, most delicate adjustment to how parameters are passed to the various PostgreSQL drivers invariably leads to dozens of regressions reported, so 100% no way would I turn this on by default.       

I would prefer if psycopg's SQL caching could be more intelligent here, such as parsing for large statements with IN() constructs.  What does psycopg do for a large INSERT statement with many VALUES expressions, does that blow out the cache also?  Because we send statements like that as well.

**Reply by CaselIT** [reply]
> In my examples, the textual queries take 12KB to 70KB (70 * 128 => 9MB) but the parsed query cache takes more than 100MB, there is an order of magnitude between the textual queries and psycopg parsed queries.

casually looking at the the code posted by @dvarrazzo https://github.com/psycopg/psycopg/blob/b8f27fc81be8e681a1cf11c36926dc16f98c7243/psycopg/psycopg/_queries.py#L147 maybe a better way (but not as automatic) to reduce the size used would be to cache the hash, not the compete input query.

Currently the cache is `2 x query_size_str + bunch of other things`. Depending on how big is the query_size_str compared to the total size, caching the hash of the original query, going to `hash original + query_size_str + buch of other things`, may have substantial size reduction?

**Reply by dvarrazzo** [reply]
> In my examples, the textual queries take 12KB to 70KB (70 * 128 => 9MB) but the parsed query cache takes more than 100MB, there is an order of magnitude between the textual queries and psycopg parsed queries.

The cache caches the result of a complex transformation:

```python
>>> psycopg._queries._query2pg(b"select %(foo)s, %(bar)s, %(foo)s", "utf8")
(b'select $1, $2, $1',
 [<PyFormat.AUTO: 's'>, <PyFormat.AUTO: 's'>],
 ['foo', 'bar'],
 [QueryPart(pre=b'select ', item='foo', format=<PyFormat.AUTO: 's'>),
  QueryPart(pre=b', ', item='bar', format=<PyFormat.AUTO: 's'>),
  QueryPart(pre=b', ', item='foo', format=<PyFormat.AUTO: 's'>),
  QueryPart(pre=b'', item=0, format=<PyFormat.AUTO: 's'>)])
```

the query is just the key, the value is something proportional to its size and its number of parameter.

This is an implementation detail: someone was also working at rewriting this operation in C so it might even be that the cache will not be needed anymore, or that the footprint will decrease. The presence of a cache here is _really_ an implementation detail.

**Comment by zzzeek** [comment]
this cache is apparently *global*.   non-public, can't be accessed, so if an application fills up 1G of data in the cache, *closes all psycopg connections*, their app is permanently stuck with 1G of unused data?   

this cache should **really** be exposed in some way, some way of clearing it, controlling its size, controlling its use, controlling the span of what statements are cached and how long they last.  This is simply standard practice when you have a cache.

Asyncpg uses caching and absolutely exposes this; at https://magicstack.github.io/asyncpg/current/api/index.html#connection  see `statement_cache_size=100`, `max_cached_statement_lifetime=300`, `max_cacheable_statement_size=15360`.  These are basic settings anytime you are unconditionally putting SQL strings in caches.

SQLAlchemy's own cache is of course fully configurable as well, https://docs.sqlalchemy.org/en/20/core/connections.html#sql-compilation-caching where you can disable it, set its size, use different caches per statement, etc.

PostgreSQL itself caches things on sessions.   It therefore also has management options including DISCARD, RESET, CLOSE ALL, etc., 

On our end, using ANY for IN() is "fine", but this is a band aid.   We send INSERT statements like this:

     INSERT INTO table VALUES (%s, %s, %s, %s), (%s, %s, %s, %s), (%s, %s, %s, %s), (%s, %s, %s, %s), ...

These are also going to blow out this cache and a lot faster.      

If psycopg3 is really unconditionally caching SQL strings **globally** with no intelligent parsing of the SQL or any way of exposing this system so that the cache size can be adjusted, used per connection, etc, we will be getting bug reports like this over, and over, and over again.       it absolutely **must** expose this cache to be able to control for large statements.

**Reply by dvarrazzo** [reply]
> this cache is apparently global. non-public, can't be accessed, so if an application fills up 1G of data in the cache, closes all psycopg connections, their app is permanently stuck with 1G of unused data?

You need to write a pathological program to do so, otherwise, caching the parsing of queries shouldn't be a problem.

> Asyncpg uses caching and absolutely exposes this

These are controls for prepared statements, not for plain parsing of a string. There is no need to establish a lifetime for a parsing result (not that I see the need of a lifetime for a prepared statement).

> SQLAlchemy's own cache is of course fully configurable as well

I assume people use the default parameters, because there is no reason for a person to disable a cache that a library maintainer has put there for a reason without a need to do so. People might want to fudge the parameters once they have found a problem... or they can fix what they have found to be problematic. Both require manual intervention of the author of the program, but for me fudging the parameters is a more band-aid-solution than noticing that a query is inefficient and writing it better.

> We send INSERT statements like this:
>
> INSERT INTO table VALUES (%s, %s, %s, %s), (%s, %s, %s, %s), (%s, %s, %s, %s), (%s, %s, %s, %s), ...
>
> These are also going to blow out this cache and a lot faster.

COPY, or executemany, could be used for better efficiency. In this case, an executemany entering a single record would use automatically a prepared statement and pipeline mode, instead of bloating the parser (not only the client's but the server's too).

> [psycopg 3] absolutely must expose this cache to be able to control for large statements.

We can disable caching for silly large number of parameters (> 50?) or silly large statements (> 4k).

**Comment by zzzeek** [comment]
> > We send INSERT statements like this:
> > INSERT INTO table VALUES (%s, %s, %s, %s), (%s, %s, %s, %s), (%s, %s, %s, %s), (%s, %s, %s, %s), ...
> > These are also going to blow out this cache and a lot faster.
> 
> COPY, or executemany, could be used for better efficiency. In this case, an executemany entering a single record would use automatically a prepared statement and pipeline mode, instead of bloating the parser (not only the client's but the server's too).

COPY is nothing like an INSERT (SQL expressions, triggers(?), CTEs, ON DUPLICATE KEY, etc) and we also use RETURNING with these statements, which is why COPY / executemany are not options.  that said, our INSERT batching feature is also fully configurable (see https://docs.sqlalchemy.org/en/20/core/connections.html#insert-many-values-behavior-for-insert-statements)


> 
> > [psycopg 3] absolutely must expose this cache to be able to control for large statements.
> 
> We can disable caching for silly large number of parameters (> 50?) or silly large statements (> 4k).

I think that's the way to go here.

My only point is that this is the state of things, and if psycopg3 appears to be leaking memory to end users when they run lots of INSERT statements with thousands of rows in the VALUES clause, these reports will keep coming in.

**Reply by dvarrazzo** [reply]
> COPY is nothing like an INSERT (SQL expressions, triggers(?), CTEs, ON DUPLICATE KEY, etc) and we also use RETURNING with these statements, which is why COPY / executemany are not options.

Unrelated, but in psycopg 3.1 executemany supports returning, in case you may want to make use of it.

>> We can disable caching for silly large number of parameters (> 50?) or silly large statements (> 4k).

> I think that's the way to go here.

I think we can provide it in the next 3.1.11, as it wouldn't change any public interface.

**Reply by zzzeek** [reply]
our INSERT..(MANY)...RETURNING was inspired by psycopg2's feature.   In practice it does a lot more.  if the psycopg3 feature is only doing the same rewrite that we already are, we are better sticking with our current approach so that we dont have a totally different implementation with different behavioral artifacts.

in particular, note that INSERT...VALUES..RETURNING does **not** guarantee the order of the VALUES parameters would be aligned with the rows from RETURNING.  Does psycopg3's feature take this into account?  The rows RETURNED can be in any order compared to the order in which the parameter sets were sent.  SQLAlchemy provides the option to enforce that these two orderings match, which we do by using a **much** more complicated syntax that we spent many weeks working on, checking with PG developers including Tom Lane, etc.

**Reply by dvarrazzo** [reply]
A `INSERT ... VALUES (one record) RETURNING ...` into an executemany with a sequence of N entries results in N queries, but prepared and pipelined. Details about how it looks on the wire are [here](https://github.com/psycopg/psycopg/pull/260#issuecomment-1080649952). The values returned are in the same order of the queries executed because each entry would be its own query.

It might still not be a win compared to the economies of scale that might happen with an `INSERT ... VALUES (many records)`, and I don't know about what other care you have to take in order to ensure that input and output orders match, so it's just an FYI in case you want to look into it more.

**Reply by zzzeek** [reply]
OK, then that is more like what cx_oracle and oracledb do; we do use their executemany() feature since they have a completely custom way of getting the RETURNING (which are more like RETURNING INTO <outparams>)

**Reply by CaselIT** [reply]
> A INSERT ... VALUES (one record) RETURNING ... into an executemany with a sequence of N entries results in N queries, but prepared and pipelined. Details about how it looks on the wire are https://github.com/psycopg/psycopg/pull/260#issuecomment-1080649952. The values returned are in the same order of the queries executed because each entry would be its own query.

> It might still not be a win compared to the economies of scale that might happen with an INSERT ... VALUES (many records), and I don't know about what other care you have to take in order to ensure that input and output orders match, so it's just an FYI in case you want to look into it more.

does this mode need to do several round networks round trips (usually one per values)? If so it's very likely to perform worse because of the network latency, even if the server time may well be less with this approach

**Reply by dvarrazzo** [reply]
> does this mode need to do several round networks round trips (usually one per values)? 

no, in pipeline mode first all the queries are sent, then all values are returned (unless some buffer is full and forces to reverse the flow).

**Reply by CaselIT** [reply]
Interesting. does it require some special flag, or you just need to do `executemany`? 

Assuming it performs better @zzzeek do you think it's worth the effort to have psycopg3 do something different than the other pg dialect? I guess we have the oracle example so the code should already support a dialect that does executemany and has no need for flags etc to keep order 

(sorry this is a bit OT)

**Reply by dvarrazzo** [reply]
> Interesting. does it require some special flag, or you just need to do executemany?

`executemany()` uses the pipeline mode automatically, as long as the client has libpq v14 or newer installed. Of course `psycopg-binary` packages a libpq modern enough. In more general cases pipeline mode must be enabled manually, which, syntactically speaking, is done via a context manager block.

Docs at https://www.psycopg.org/psycopg3/docs/advanced/pipeline.html

**Reply by CaselIT** [reply]
@dvarrazzo a couple of comments on that doc page. 

Would be useful to add an executemamy with an insert-returning to the examples.

It's also not clear to me if multiple retuning queries are supported, like two unrelated selects, and how would that behave to fetch rhe results

**Reply by dvarrazzo** [reply]
Thank you for the feedback, @CaselIT 

The detail you ask is documented in the [`executemany()` docs](https://www.psycopg.org/psycopg3/docs/api/cursors.html#psycopg.Cursor.executemany) (you can use `nextset()` to access several results from executemany).

We can think about some docs reshuffling to put other related topics together.

**Reply by CaselIT** [reply]
(btw done a test on execute many vs single insert with many values and the single statement is still a bit faster, but not by much. great work!)
<details><summary>Details</summary>
<p>

```py

import time
import psycopg

url = "postgresql://scott:tiger@127.0.0.1:5432/test"
conn = psycopg.connect(url)

conn.execute(
    """
create table tbl(
  id serial primary key,
  a int,
  b text,
  c double precision,
  d int
)"""
)

conn.commit()

data = [(i, f"value {i}", i / 42, -i) for i in range(10000)]


def gomany():
    with psycopg.connect(url) as c:
        for i in range(5):
            c.execute("select 42").fetchall()

        with c.cursor() as cur:
            start = time.time()
            cur.executemany(
                "insert into tbl(a,b,c,d) values (%s, %s, %s, %s)", data
            )
            end = time.time()
            cur.close()
        print("execute many time", end - start)


def truncate():
    conn.execute("truncate table tbl")
    conn.commit()


def gosingle():
    val = ",".join(["(%s, %s, %s, %s)"] * len(data))

    single_val = sum(data, ())

    stmt = "insert into tbl(a,b,c,d) values " + val
    with psycopg.connect(url) as c:
        for i in range(5):
            c.execute("select 42").fetchall()

        with c.cursor() as cur:
            start = time.time()
            cur.execute(stmt, single_val)
            end = time.time()
            cur.close()
        print("single insert val ues time", end - start)

for _ in range(5):
    truncate()
    gomany()
    truncate()
    gosingle()

conn.execute("drop table tbl")
conn.commit()
conn.close()
```
result (on localhost)
```
execute many time 0.1727004051208496
single insert val ues time 0.12656354904174805
execute many time 0.17663884162902832
single insert val ues time 0.06110715866088867
execute many time 0.17415857315063477
single insert val ues time 0.06152009963989258
execute many time 0.1775827407836914
single insert val ues time 0.06652951240539551
execute many time 0.18039631843566895
single insert val ues time 0.06252717971801758
```

</p>
</details>
